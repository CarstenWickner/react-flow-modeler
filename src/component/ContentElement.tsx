import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export class ContentElement extends React.Component<{
    elementId: string;
    editMenu: (() => React.ReactNode) | undefined;
    onSelect: (elementId: string) => void;
}> {
    onClick = (event: React.MouseEvent): void => {
        const { onSelect, elementId } = this.props;
        onSelect(elementId);
        event.stopPropagation();
    };

    render(): React.ReactNode {
        const { editMenu, children } = this.props;
        return (
            <>
                <div className="stroke-horizontal arrow" />
                <div className="flow-element-wrapper">
                    <div className={`flow-element content-element${editMenu ? " selected" : ""}`} onClick={this.onClick}>
                        {children}
                    </div>
                    {editMenu && editMenu()}
                </div>
                <HorizontalStroke optional />
            </>
        );
    }
}
