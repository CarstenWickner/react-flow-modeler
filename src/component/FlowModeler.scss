@import "node_modules/mathsass/dist/math";

.flow-modeler {
    display: grid;
    padding: 2em 1em;
    align-content: stretch;
    justify-content: stretch;

    $strokeColor: #01579b;
    $dropStrokeColor: #00b0ff;
    $strokeStrength: 0.25em;
    $arrowSize: 1.5 * $strokeStrength;
    $strokeLength: 1.75em;
    $elementSizeFactor: 2.5;
    $standardElementSize: $elementSizeFactor * 1em;
    $gatewayBorderSize: sqrt($elementSizeFactor * $elementSizeFactor * 2) * 1em / 2;
    $standardElementColor: #808080;
    $selectionColor: #ffab00;

    & .grid-cell {
        display: flex;
        align-items: center;
        align-content: center;
        min-height: $strokeLength;
        position: relative;
    }
    & .flow-element-wrapper {
        display: flex;
        align-items: center;
        position: relative;
    }

    & .flow-element {
        display: flex;
        position: relative;
        margin: $strokeStrength 0;

        &.start-element {
            border: ($strokeStrength / 2) solid $standardElementColor;
            height: $standardElementSize - $strokeStrength;
            width: $standardElementSize - $strokeStrength;
            border-radius: 50%;
        }
        &.end-element {
            box-shadow: 0 0 0 ($strokeStrength * 0.9) $standardElementColor;
            border: ($strokeStrength * 0.9) solid white;
            background-color: $standardElementColor;
            height: $standardElementSize - (4 * 0.9 * $strokeStrength);
            width: $standardElementSize - (4 * 0.9 * $strokeStrength);
            border-radius: 50%;
            margin: $strokeStrength ($strokeStrength * 0.9);
        }
        &.step-element {
            border: ($strokeStrength / 3) solid $standardElementColor;
            border-radius: 0.25em;
            min-width: $standardElementSize;
            min-height: $standardElementSize / 2;
        }
        &.gateway-element {
            position: relative;
            height: $standardElementSize;
            width: $standardElementSize;
            margin: 0 $strokeStrength / 2;
            z-index: 1;

            display: flex;
            align-items: center;
            justify-content: center;

            &::before {
                content: "";
                position: absolute;
                transform: rotate(45deg);
                border: ($strokeStrength / 2) solid $standardElementColor;
                height: $gatewayBorderSize;
                width: $gatewayBorderSize;
                background-color: transparent;
            }
            &::after {
                content: "X";
                position: absolute;
                font: bold 1.75em Arial, sans-serif;
                color: $standardElementColor;
            }
        }
        &.start-element.selected,
        &.step-element.selected,
        &.gateway-element.selected::before {
            border-color: $selectionColor;
        }
    }
    & .menu {
        position: absolute;
        display: flex;
        flex-direction: column;
        align-items: center;
        left: calc(100% + #{$strokeStrength});
        width: 1em;
        margin: auto;
        border: ($strokeStrength / 3) solid $selectionColor;
        border-radius: 0.25em;
        background-color: white;
        box-shadow: 0.1em 0.1em 0.25em 0.25em rgba(0, 0, 0, 0.25);
        z-index: 1;
        padding: $strokeStrength;

        &-item {
            font-family: Arial, Helvetica, sans-serif;
            line-height: 1.1em;
            cursor: pointer;
    
            &.add-branch::before {
                content: "↳";
            }
            &.add-step::before {
                content: "■";
            }
            &.add-gateway::before {
                content: "◆";
            }
            &.change-next::before {
                content: "⇥";
            }
            &.remove::before {
                content: "×";
            }
        }
    }
    & .gateway-element.diverging + .menu {
        top: ($strokeStrength * 3) + (1.1em / 2);
    }
    & .stroke-horizontal::before,
    & .stroke-vertical::before {
        background-color: $strokeColor;
    }
    &.editable .stroke-horizontal.selected::before {
        background-color: $selectionColor;
    }
    & .stroke-horizontal {
        position: relative;
        display: flex;
        flex-direction: column;
        flex-grow: 1;
        min-width: $strokeLength;

        &.optional {
            min-width: 0;
        }
        &.arrow {
            margin-right: $arrowSize;
    
            &::after {
                content: "";
                height: 0;
                width: 0;
                position: absolute;
                right: 0;
                top: calc(50% - #{$arrowSize});
                margin-right: -$arrowSize;
                border-top: $arrowSize solid transparent;
                border-left: $arrowSize solid $strokeColor;
                border-bottom: $arrowSize solid transparent;
            }
            &.can-drop::after {
                border-left-color: $dropStrokeColor;
            }
        }
        &::before {
            position: absolute;
            content: "";
            height: $strokeStrength;
            width: 100%;
            top: calc(50% - #{$strokeStrength / 2});
        }
        &.can-drop::before {
            background: linear-gradient(90deg, $strokeColor 0%, $dropStrokeColor 62%);
        }
        & > .top-label {
            display: flex;
            flex-direction: column;
            justify-content: flex-end;
            align-content: center;
            padding: 0 0.25em;
            margin-bottom: $strokeStrength / 2;
            min-height: $strokeStrength * 4;
        }
        & > .bottom-spacing {
            display: flex;
            flex-direction: column;
            margin-bottom: $strokeStrength / 2;
        }
        & .clickable-spacing {
            min-height: $strokeStrength * 4;
            max-height: $strokeStrength * 8;
            width: 100%;
        }
    }
    & .stroke-vertical {
        position: relative;
        display: flex;
        flex-direction: column;
        align-self: stretch;
        width: $strokeStrength;
        &::before {
            position: absolute;
            content: "";
            width: $strokeStrength;
        }
        &.top-half::before {
            height: calc(50% + #{$strokeStrength / 2});
            top: 0;
        }
        &.full-height::before {
            height: 100%;
            top: 0;
        }
        &.bottom-half::before {
            height: calc(50% + #{$strokeStrength / 2});
            bottom: 0;
        }
    }
    &.editable {
        & .flow-element,
        & .top-label,
        & .clickable-spacing {
            cursor: context-menu;
        }
        & .flow-element.end-element,
        & .selected.flow-element,
        & .selected .top-label,
        & .selected .clickable-spacing {
            cursor: inherit;
        }
        & .start-element.selected,
        & .start-element:hover,
        & .step-element.selected,
        & .step-element:hover,
        & .gateway-element.selected::before,
        & .gateway-element:hover::before {
            box-shadow: 0 0 0.1em 0.1em rgba(0, 0, 0, 0.25);
        }
        & .flow-element-wrapper.can-drop > .step-element,
        & .flow-element-wrapper.can-drop > .gateway-element::before {
            box-shadow: 0 0 0.125em 0.125em rgba(0, 176, 255, 0.5);
        }
        & .flow-element-wrapper.can-drop > .end-element {
            box-shadow:
                0 0 0 ($strokeStrength * 0.9) $standardElementColor,
                0 0 0.125em ($strokeStrength * 0.9 + 0.125em) rgba(0, 176, 255, 0.5);
        }
        & .dragging {
            color: $dropStrokeColor;
        }
        & .flow-element-wrapper.cannot-drop,
        & .flow-element-wrapper.cannot-drop > .flow-element {
            cursor: no-drop;
        }
    }
}
