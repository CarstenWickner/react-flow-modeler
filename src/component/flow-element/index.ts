export { Start } from "./Start";
export { ContentElement } from "./ContentElement";
export { Gateway } from "./Gateway";
export { GatewayConnector } from "./GatewayConnector";
export { StrokeExtension } from "./StrokeExtension";
export { End } from "./End";
