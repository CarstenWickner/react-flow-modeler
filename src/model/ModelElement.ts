export const enum ElementType {
    Start = 1,
    Content,
    GatewayDiverging,
    GatewayConverging,
    ConnectGatewayToElement,
    ConnectElementToGateway,
    End
}

interface BaseModelNode {
    columnIndex: number;
    rowCount: number;
}

export interface StartNode extends BaseModelNode {
    type: ElementType.Start;
    followingElement: ContentNode | DivergingGatewayNode | EndNode;
}

export interface ContentNode extends BaseModelNode {
    type: ElementType.Content;
    id: string;
    precedingElement: StartNode | ContentNode | DivergingGatewayBranch | ConvergingGatewayNode;
    followingElement: ContentNode | DivergingGatewayNode | ConvergingGatewayBranch | EndNode;
    data?: { [key: string]: unknown };
}

export interface DivergingGatewayNode extends BaseModelNode {
    type: ElementType.GatewayDiverging;
    id: string;
    precedingElement: StartNode | ContentNode | DivergingGatewayBranch | ConvergingGatewayNode;
    followingBranches: Array<DivergingGatewayBranch>;
    data?: { [key: string]: unknown };
}

export interface DivergingGatewayBranch extends BaseModelNode {
    type: ElementType.ConnectGatewayToElement;
    precedingElement: DivergingGatewayNode;
    branchIndex: number;
    followingElement: ContentNode | DivergingGatewayNode | ConvergingGatewayBranch;
    data?: { [key: string]: unknown };
}

export interface ConvergingGatewayNode extends BaseModelNode {
    type: ElementType.GatewayConverging;
    precedingBranches: Array<ConvergingGatewayBranch>;
    followingElement: ContentNode | DivergingGatewayNode | EndNode;
}

export interface ConvergingGatewayBranch extends BaseModelNode {
    type: ElementType.ConnectElementToGateway;
    precedingElement: ContentNode | DivergingGatewayBranch;
    followingElement: ConvergingGatewayNode;
    branchIndex: number;
}

export interface EndNode extends BaseModelNode {
    type: ElementType.End;
    precedingElement: StartNode | ContentNode | ConvergingGatewayNode;
}

export type ModelElementExclStart =
    | ContentNode
    | DivergingGatewayNode
    | DivergingGatewayBranch
    | ConvergingGatewayBranch
    | ConvergingGatewayNode
    | EndNode;

export type ModelElement = StartNode | ModelElementExclStart;
