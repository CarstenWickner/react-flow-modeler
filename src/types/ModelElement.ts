export const enum ElementType {
    StartNode = "start",
    StepNode = "step",
    DivergingGatewayNode = "div-gw",
    ConvergingGatewayNode = "conv-gw",
    DivergingGatewayBranch = "div-branch",
    ConvergingGatewayBranch = "conv-branch",
    EndNode = "end"
}

interface BaseModelNode {
    columnIndex: number;
    rowCount: number;
}

export interface StartNode extends BaseModelNode {
    type: ElementType.StartNode;
    followingElement: StepNode | DivergingGatewayNode | EndNode;
}

export interface StepNode extends BaseModelNode {
    type: ElementType.StepNode;
    id: string;
    precedingElement: StartNode | StepNode | DivergingGatewayBranch | ConvergingGatewayNode;
    followingElement: StepNode | DivergingGatewayNode | ConvergingGatewayBranch | EndNode;
    data?: { [key: string]: unknown };
}

export interface DivergingGatewayNode extends BaseModelNode {
    type: ElementType.DivergingGatewayNode;
    id: string;
    precedingElement: StartNode | StepNode | DivergingGatewayBranch | ConvergingGatewayNode;
    followingBranches: Array<DivergingGatewayBranch>;
    data?: { [key: string]: unknown };
}

export interface DivergingGatewayBranch extends BaseModelNode {
    type: ElementType.DivergingGatewayBranch;
    precedingElement: DivergingGatewayNode;
    branchIndex: number;
    followingElement: StepNode | DivergingGatewayNode | ConvergingGatewayBranch;
    data?: { [key: string]: unknown };
}

export interface ConvergingGatewayNode extends BaseModelNode {
    type: ElementType.ConvergingGatewayNode;
    precedingBranches: Array<ConvergingGatewayBranch>;
    followingElement: StepNode | DivergingGatewayNode | EndNode;
}

export interface ConvergingGatewayBranch extends BaseModelNode {
    type: ElementType.ConvergingGatewayBranch;
    precedingElement: StepNode | DivergingGatewayBranch;
    followingElement: ConvergingGatewayNode;
    branchIndex: number;
}

export interface EndNode extends BaseModelNode {
    type: ElementType.EndNode;
    precedingElement: StartNode | StepNode | ConvergingGatewayNode;
}

export type ModelElementExclStart =
    | StepNode
    | DivergingGatewayNode
    | DivergingGatewayBranch
    | ConvergingGatewayBranch
    | ConvergingGatewayNode
    | EndNode;

export type ModelElement = StartNode | ModelElementExclStart;
