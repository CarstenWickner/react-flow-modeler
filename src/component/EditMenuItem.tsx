import * as React from "react";
import { useDrag } from "react-dnd";

import { DraggedLinkContext } from "../types/EditAction";

const disabledDragging: [{ isDragging: boolean }, React.LegacyRef<HTMLSpanElement>] = [{ isDragging: false }, undefined];

export const EditMenuItem: React.FC<{
    options: { className?: string; title?: string };
    defaultClassName: string;
    onClick?: (event: React.MouseEvent) => void;
    dragItem?: DraggedLinkContext;
}> = ({ options, defaultClassName, onClick, dragItem }) => {
    if (!dragItem && !onClick) {
        return null;
    }
    const [{ isDragging }, drag] = dragItem
        ? useDrag({
              item: dragItem,
              collect: (monitor) => ({ isDragging: !!monitor.isDragging() })
          })
        : disabledDragging;
    const baseClassName = (options && options.className) || `menu-item ${defaultClassName}`;
    return (
        <span
            className={`${baseClassName}${isDragging ? " dragging" : ""}`}
            {...{ title: options ? options.title : undefined, ref: drag, onClick }}
        />
    );
};
