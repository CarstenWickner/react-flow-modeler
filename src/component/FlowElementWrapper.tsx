import * as React from "react";
import { useDrop } from "react-dnd";

import { isFlowValid } from "../model/pathValidationUtils";

import { ContentNode, ConvergingGatewayNode, DivergingGatewayNode, ElementType, EndNode, ModelElementExclStart } from "../types/ModelElement";
import { DraggableType, DraggedLinkContext, onLinkDropCallback } from "../types/EditAction";

const isTargetAncestorOfDragItem = (
    originElement: ModelElementExclStart,
    referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode
): boolean => {
    if (originElement === referenceElement) {
        return true;
    }
    if (originElement.type === ElementType.ConvergingGatewayNode) {
        return originElement.precedingBranches.some((branch) => isTargetAncestorOfDragItem(branch, referenceElement));
    }
    return (
        originElement.precedingElement.type !== ElementType.StartNode && isTargetAncestorOfDragItem(originElement.precedingElement, referenceElement)
    );
};

const isDropValid = (referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode, onDrop: onLinkDropCallback) => (
    dragContext: DraggedLinkContext
): boolean =>
    !isTargetAncestorOfDragItem(dragContext.originElement, referenceElement) &&
    dragContext.originElement.followingElement !== referenceElement &&
    isFlowValid(onDrop(referenceElement, dragContext, true).changedFlow);

const disabledDropping: [{ isOver: boolean; canDrop: false }, React.LegacyRef<HTMLDivElement>] = [{ isOver: false, canDrop: false }, undefined];

export const FlowElementWrapper: React.FC<{
    elementTypeClassName: string;
    referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode;
    editMenu?: (() => React.ReactNode) | undefined;
    onLinkDrop?: onLinkDropCallback | undefined;
    onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
}> = ({ elementTypeClassName, referenceElement, editMenu, onLinkDrop, onClick, children }) => {
    const [{ isOver, canDrop }, drop] = onLinkDrop
        ? useDrop({
              accept: DraggableType.LINK,
              canDrop: isDropValid(referenceElement, onLinkDrop),
              drop: (dragContext: DraggedLinkContext) => onLinkDrop(referenceElement, dragContext),
              collect: (monitor) => ({
                  isOver: !!monitor.isOver(),
                  canDrop: !!monitor.canDrop()
              })
          })
        : disabledDropping;
    return (
        <>
            <div className={`stroke-horizontal arrow${isOver && canDrop ? " can-drop" : ""}`} />
            <div className={`flow-element-wrapper${isOver ? (canDrop ? " can-drop" : " cannot-drop") : ""}`} ref={drop}>
                <div className={`flow-element ${elementTypeClassName}${editMenu ? " selected" : ""}`} onClick={onClick}>
                    {children}
                </div>
                {editMenu && editMenu()}
            </div>
        </>
    );
};
