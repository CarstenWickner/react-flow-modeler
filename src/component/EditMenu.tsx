import * as React from "react";

import { EditMenuItem } from "./EditMenuItem";
import { FlowElementReference } from "../model/FlowElement";
import { addContentElement, addDivergingGateway } from "../model/action/addElement";
import { addBranch } from "../model/action/addBranch";
import { isChangeNextElementAllowed } from "../model/action/changeNextElement";
import { removeElement, isRemoveElementAllowed } from "../model/action/removeElement";

import { SelectableElementType, EditActionResult, DraggableType } from "../types/EditAction";
import { FlowModelerProps, MenuOptions } from "../types/FlowModelerProps";
import { ElementType } from "../types/GridCellData";

const onClickStopPropagation = (event: React.MouseEvent): void => event.stopPropagation();

export class EditMenu extends React.Component<{
    targetType: SelectableElementType;
    referenceElement?: FlowElementReference;
    branchIndex?: number;
    menuOptions?: FlowModelerProps["options"]["editActions"];
    onChange: (change: (originalFlow: FlowModelerProps["flow"]) => EditActionResult) => void;
}> {
    isNextElementReferencedByOthers = (): boolean => {
        const { referenceElement, branchIndex } = this.props;
        if (!referenceElement) {
            return false;
        }
        return referenceElement.getFollowingElements()[branchIndex || 0].getPrecedingElements().length > 1;
    };

    renderMenuItem(
        options: MenuOptions,
        defaultClassName: string,
        onClick?: (event: React.MouseEvent) => void,
        dragType?: DraggableType
    ): React.ReactNode {
        const { targetType, referenceElement, branchIndex } = this.props;
        if (options && options.isActionAllowed && !options.isActionAllowed(targetType, referenceElement, branchIndex)) {
            return null;
        }
        const dragItem =
            dragType === DraggableType.LINK &&
            (targetType === ElementType.Content || targetType == ElementType.ConnectGatewayToElement) &&
            this.isNextElementReferencedByOthers()
                ? { type: dragType, originType: targetType, originElement: referenceElement, originBranchIndex: branchIndex }
                : undefined;
        return <EditMenuItem key={defaultClassName} options={options} defaultClassName={defaultClassName} onClick={onClick} dragItem={dragItem} />;
    }

    onAddContentElementClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            onChange((originalFlow) => addContentElement(originalFlow, targetType, {}, referenceElement, branchIndex));
        }
    };

    renderAddContentElementItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(menuOptions ? menuOptions.addFollowingContentElement : undefined, "add-content", this.onAddContentElementClick);
    }

    onAddDivergingGatewayClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            onChange((originalFlow) => addDivergingGateway(originalFlow, targetType, {}, referenceElement, branchIndex));
        }
    };

    renderAddDivergingGatewayItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(
            menuOptions ? menuOptions.addFollowingDivergingGateway : undefined,
            "add-gateway",
            this.onAddDivergingGatewayClick
        );
    }

    onAddDivergingBranchClick = (): void => {
        const { targetType, onChange, referenceElement } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            onChange((originalFlow) => addBranch(originalFlow, {}, referenceElement));
        }
    };

    renderAddDivergingBranchItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(menuOptions ? menuOptions.addDivergingBranch : undefined, "add-branch", this.onAddDivergingBranchClick);
    }

    renderChangeNextElementItem(): React.ReactNode {
        const { targetType, referenceElement, branchIndex, menuOptions } = this.props;
        if (isChangeNextElementAllowed(targetType, referenceElement, branchIndex)) {
            return this.renderMenuItem(menuOptions ? menuOptions.changeNextElement : undefined, "change-next", undefined, DraggableType.LINK);
        }
        return null;
    }

    onRemoveClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType === ElementType.Content || targetType === ElementType.ConnectGatewayToElement) {
            onChange((originalFlow) => removeElement(originalFlow, targetType, referenceElement, branchIndex));
        }
    };

    renderRemoveItem(): React.ReactNode {
        const { targetType, referenceElement, branchIndex, menuOptions } = this.props;
        if (isRemoveElementAllowed(targetType, referenceElement, branchIndex)) {
            return this.renderMenuItem(menuOptions ? menuOptions.removeElement : undefined, "remove", this.onRemoveClick);
        }
        return null;
    }

    render(): React.ReactNode {
        const menuItems = [
            this.renderAddContentElementItem(),
            this.renderAddDivergingGatewayItem(),
            this.renderAddDivergingBranchItem(),
            this.renderChangeNextElementItem(),
            this.renderRemoveItem()
        ];
        if (menuItems.some((item) => item)) {
            return (
                <div className="menu" onClick={onClickStopPropagation}>
                    {menuItems}
                </div>
            );
        }
        return null;
    }
}
