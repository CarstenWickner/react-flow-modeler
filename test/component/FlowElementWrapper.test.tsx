import * as React from "react";
import { shallow } from "enzyme";

import { FlowElementWrapper } from "../../src/component/FlowElementWrapper";
import { ContentNode, ElementType } from "../../src/types/ModelElement";

const mockFlowElementReference = (id: string): ContentNode => ({
    type: ElementType.ContentNode,
    id,
    precedingElement: undefined,
    followingElement: undefined,
    columnIndex: 2,
    rowCount: 1
});

describe("renders correctly", () => {
    const onSelect = (): void => {};
    it("with minimal props", () => {
        const component = shallow(
            <FlowElementWrapper
                elementTypeClassName="some-element"
                referenceElement={mockFlowElementReference("element-id")}
                editMenu={(): React.ReactNode => <div className="edit-menu-placeholder" />}
                onLinkDrop={undefined}
                onSelect={onSelect}
            >
                {"text"}
            </FlowElementWrapper>
        );
        expect(component).toMatchSnapshot();
    });
    it("when not selected", () => {
        const component = shallow(
            <FlowElementWrapper
                elementTypeClassName="some-element"
                referenceElement={mockFlowElementReference("element-id")}
                editMenu={undefined}
                onLinkDrop={undefined}
                onSelect={onSelect}
            >
                {"text"}
            </FlowElementWrapper>
        );
        const contentElement = component.find(".some-element");
        expect(contentElement.hasClass("selected")).toBe(false);
        expect(component.find(".edit-menu-placeholder").exists()).toBe(false);
    });
});
