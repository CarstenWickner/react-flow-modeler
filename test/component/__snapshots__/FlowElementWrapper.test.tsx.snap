// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal props 1`] = `
<Fragment>
  <div
    className="stroke-horizontal arrow"
  />
  <div
    className="flow-element-wrapper"
  >
    <div
      className="flow-element some-element selected"
    >
      text
    </div>
    <div
      className="edit-menu-placeholder"
    />
  </div>
</Fragment>
`;
